reconstructSPI/iterative_refinement/expectation_maximization.py

```diff
@@ -0,0 +1,539 @@
+        # space the whole time. They are experimental measurements
+        # and are fixed in the algorithm
+        particles_f_1 = IterativeRefinement.fft_3d(particles_1)
+        particles_f_2 = IterativeRefinement.fft_3d(particles_2)
+
+        n_pix = self.map_3d_init.shape[0]
+
+        n_rotations = self.particles.shape[0]
+
+        # suggest 32 or 64 to start with. real data will be more like
+        # 128 or 256. Can have issues with ctf at small pixels and
+        # need to zero pad to avoid artefacts. Artefacts from ctf not
+        # going to zero at edges, and sinusoidal ctf rippling too fast
+        # can zero pad when do Fourier convolution (fft is on zero
+        # padded and larger sized array)
+
+        half_map_3d_r_1, half_map_3d_r_2 = (
+            self.map_3d_init.copy(),
+            self.map_3d_init.copy(),
+        )
+        # should diverge because different particles averaging in
+
+        half_map_3d_f_1 = IterativeRefinement.fft_3d(half_map_3d_r_1)
+        half_map_3d_f_2 = IterativeRefinement.fft_3d(half_map_3d_r_2)
+
+        for _ in range(self.max_itr):
+
+            half_map_3d_f_1 = IterativeRefinement.fft_3d(half_map_3d_r_1)
+            half_map_3d_f_2 = IterativeRefinement.fft_3d(half_map_3d_r_2)
+
+            # align particles to 3D volume
+            # decide on granularity of rotations
+            # i.e. how finely the rotational SO(3) space is sampled
+            # in a grid search. Smarter method is branch and bound...
+            # perhaps can make grid up front of slices, and then only
+            # compute norms on finer grid later. So re-use slices.
+
+            # def do_adaptive_grid_search(particle, map_3d):
+            #   a la branch and bound. Not sure exactly how you decide
+            #   how finely gridded to make it. Perhaps heuristics
+            #   based on how well the signal agrees in half_map_1,
+            #   half_map_2 (Fourier frequency).
+
+            rots = IterativeRefinement.grid_SO3_uniform(n_rotations)
+
+            xy0_plane = IterativeRefinement.generate_xy_plane(n_pix)
+
+            slices_1, xyz_rotated = IterativeRefinement.generate_slices(
+                half_map_3d_f_1, xy0_plane, n_pix, rots
+            )
+            # Here rots are the same for the half maps,
+            # but could be different in general.
+            slices_2, xyz_rotated = IterativeRefinement.generate_slices(
+                half_map_3d_f_2, xy0_plane, n_pix, rots
+            )
+
+            # initialize:
+            # complex
+            map_3d_f_updated_1 = np.zeros_like(half_map_3d_f_1)
+            # complex
+            map_3d_f_updated_2 = np.zeros_like(half_map_3d_f_2)
+            # complex
+            map_3d_f_norm_1 = np.zeros_like(half_map_3d_f_1)
+            # complex
+            map_3d_f_norm_2 = np.zeros_like(half_map_3d_f_2)
+            # float/real
+            counts_3d_updated_1 = np.zeros_like(half_map_3d_r_1)
+            # float/real
```

Comment:
remove comments. complex implied in variable name with `_f` for fourier

Reply:
Also the some comments are misleading since `map_3d_f_norm_2` and `map_3d_f_norm_1` are still complex (in Fourier space).